### ISSUE TITLE
IRIS: Keithley 2001 Multimeter IOC, OPI, etc. for calorimetry project

### ISSUE THREAD

**Issue by davidkeymer** [ISISComputingGroup/IBEX#3476]
IRIS is currently using four Keithley units (model TBD) each with a single input.  These are addressed individually (presumably using four IOCs) to measure a heater voltage.  See "calorimetry" configurations on NDXIRIS for current setup.

This ticket is to provide an IOC for a Keithley 2001 Multimeter which is a single unit with up to ten input channels (0-9) and do away with the four separate units.

The heaters are connected to channels 1 to 4 inclusive.  Channels 0 and 5 use solid state relays and are not connected to the breakout box, so the IOC should provide the option to use channels 1-4 and 6-9.  The IOC will ideally need to read the active channels simultaneously, otherwise "multiplex" them i.e. switch, read, switch, read, etc.

The unit and accessories are currently in the "test area" of R3, G39.  Contact Jon Bones for further information on the hardware, Ian Silverwood and Vicki Garcia-Sakai for project details.

**Comment by Rory-Potter-Tessella** [comment]
Relevant pull request:

- [IOC](https://github.com/ISISComputingGroup/EPICS-ioc/pull/315)
- [Support module](https://github.com/ISISComputingGroup/EPICS-Keithley_2001/pull/1)
- [EPICS](https://github.com/ISISComputingGroup/EPICS/pull/133)
- [IOC Test Framework](https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/pull/122)
- [Device Emulator](https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/79)

**Comment by Rory-Potter-Tessella** [comment]
Emailed Ian, Jon and Vicky. They want to measure the power draw of the heater using 
`P = I^2 * R = V^2 /R` 
so we only need to read voltage for the device.

**Comment by ChrisM-S** [comment]
I'm not sure if this was a hidden task in this ticket but I'm guessing that "Commission the calorimeter" on the new IRIS SETUP PC needs to be noted somewhere!  As it stands the system/hardware is setup now and all that is required is that the results of this ticket and any associated testing are done on this system.  There is a need to add connection(s) from the new SETUP MOXA which is in the rack below the existing IRIS one.  Probably the easiest RS232 link (if one is needed outside the back room atall) is to run the cable from the MOXA port to the cabin patch panel and use the RJ45 socket outside by the calorimeter.

**Comment by kjwoodsISIS** [comment]
This ticket has taken considerably more than 5 points to implement.  When this ticket is completed, we should try to understand why the ticket required so much more effort than we expected, so that we don't fall into the same trap in future.  Please document your thoughts & suggestions in this ticket.

**Comment by Rory-Potter-Tessella** [comment]
Thoughts on my experiences with the Keithley 2001 ticket:

- We didn't discuss how complicated the IOC would be to implement in depth. It was mentioned that we had supported Keithley's in the past and that it would be a "normal IOC" level of complexity. Maybe in the future create tickets to investigate the complexity of an IOC before starting to create one.
- I spent two weeks looking at already developed IOCs for similar Keithley devices (inc 2000). These did not work in the end with the Keithley 2001 as they have a slightly different command set and the IOCs were not designed for my use case. Ended up contributing a fix to a bug found in the EPICS SynApps repository fixing a buffer overflow bug in their Keithley 2000/2700 IOC.
- Keithley command set is well documented, but it is difficult to work out how they work together. Lots of time was spent just working out how to combine the commands to read the buffer.
- Lots of commands have to implemented to do a simple task with a Keithley (e.g. reading a DC voltage).
- Used a sequencer to enable the dynamism required in the IOC for the ticket. This dynamism caused me lots of headaches with the sequencer and memory access problems caused by being able to change the source of truth while the sequencer was running.
- aSub records needed to parse the readings and unit.
- Added C++ unit tests to allows me to test the dynamic pieces of the code including helper functions used in the sequencer to work out which channels to scan on as well as functions used in aSub record function to parse values. Took time to get the test framework to compile/build with the EPICS build process.
